# Import libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Load Dataset

In [2]:
data_yelp = pd.read_csv('yelp_labelled.txt', sep='\t',header=None)

In [3]:
# review and sentiment
data_yelp.head()

0  1
0                           Wow... Loved this place.  1
1                                 Crust is not good.  0
2          Not tasty and the texture was just nasty.  0
3  Stopped by during the late May bank holiday of...  1
4  The selection on the menu was great and so wer...  1

> 0-Negative

> 1-Positive 

In [4]:
# Assign column names
columan_name = ['Review', 'Sentiment']
data_yelp.columns = columan_name

In [5]:
data_yelp.head()

Review  Sentiment
0                           Wow... Loved this place.          1
1                                 Crust is not good.          0
2          Not tasty and the texture was just nasty.          0
3  Stopped by during the late May bank holiday of...          1
4  The selection on the menu was great and so wer...          1

In [6]:
# 1000 rows (reviews), 2 columns (Sentiments)
data_yelp.shape

(1000, 2)

In [7]:
# review and sentiment
data_amazon = pd.read_csv('amazon_cells_labelled.txt',sep='\t',header=None)
data_amazon.head()

0  1
0  So there is no way for me to plug it in here i...  0
1                        Good case, Excellent value.  1
2                             Great for the jawbone.  1
3  Tied to charger for conversations lasting more...  0
4                                  The mic is great.  1

> 0-Negative

> 1-Positive 

In [8]:
data_amazon.columns = columan_name

In [9]:
data_amazon.head()

Review  Sentiment
0  So there is no way for me to plug it in here i...          0
1                        Good case, Excellent value.          1
2                             Great for the jawbone.          1
3  Tied to charger for conversations lasting more...          0
4                                  The mic is great.          1

In [10]:
data_amazon.shape

(1000, 2)

In [11]:
data_imdb = pd.read_csv('imdb_labelled.txt',sep='\t',header=None)
data_imdb.head()

0  1
0  A very, very, very slow-moving, aimless movie ...  0
1  Not sure who was more lost - the flat characte...  0
2  Attempting artiness with black & white and cle...  0
3       Very little music or anything to speak of.    0
4  The best scene in the movie was when Gerardo i...  1

In [12]:
data_imdb.columns = columan_name

In [13]:
data_imdb.head()

Review  Sentiment
0  A very, very, very slow-moving, aimless movie ...          0
1  Not sure who was more lost - the flat characte...          0
2  Attempting artiness with black & white and cle...          0
3       Very little music or anything to speak of.            0
4  The best scene in the movie was when Gerardo i...          1

In [14]:
data_imdb.shape

(748, 2)

## Append all the data in a single dataframe

In [15]:
# ignore_index = True --> Because the index is a string 
data = data_yelp.append([data_amazon, data_imdb],ignore_index=True)

In [16]:
data.head()

Review  Sentiment
0                           Wow... Loved this place.          1
1                                 Crust is not good.          0
2          Not tasty and the texture was just nasty.          0
3  Stopped by during the late May bank holiday of...          1
4  The selection on the menu was great and so wer...          1

In [17]:
data.shape

(2748, 2)

### check distribution of sentiments

In [18]:
data['Sentiment'].value_counts()

1    1386
0    1362
Name: Sentiment, dtype: int64

> 1386 positive reviews

> 1362 Negative reviews

### check for null values

In [19]:
data.isnull().sum()

Review       0
Sentiment    0
dtype: int64

> no null values in the data

In [20]:
x = data['Review']
y = data['Sentiment']

# Data Cleaning

> here we will remove stopwords, punctuations

> as well as we will apply lemmatization

# Create a function to clean the data

In [22]:
import string
punct = string.punctuation
punct

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [26]:
# install spacy for stopword
!pip3 install spacy

     |████████████████████████████████| 6.2 MB 5.2 MB/s eta 0:00:01
     |████████████████████████████████| 5.8 MB 4.2 MB/s eta 0:00:01     |██████████████▍                 | 2.6 MB 3.3 MB/s eta 0:00:01
     |████████████████████████████████| 609 kB 5.6 MB/s eta 0:00:01
     |████████████████████████████████| 450 kB 5.0 MB/s eta 0:00:01
     |████████████████████████████████| 2.6 MB 5.5 MB/s eta 0:00:01
     |████████████████████████████████| 106 kB 4.8 MB/s eta 0:00:01
     |████████████████████████████████| 42 kB 2.3 MB/s eta 0:00:011
     |████████████████████████████████| 181 kB 2.9 MB/s eta 0:00:01
     |████████████████████████████████| 58 kB 3.2 MB/s eta 0:00:011
  Attempting uninstall: typing-extensions
    Found existing installation: typing-extensions 3.7.4.2
    Uninstalling typing-extensions-3.7.4.2:
      Successfully uninstalled typing-extensions-3.7.4.2
  Attempting uninstall: smart-open
    Found existing installation: smart-open 4.1.2
    Uninstalling smart-open-4.1.2:

In [27]:
from spacy.lang.en.stop_words import STOP_WORDS

In [28]:
stopwords = list(STOP_WORDS) # list of stopwords
stopwords

['‘s',
 'seemed',
 'sometime',
 'your',
 'done',
 'also',
 'nor',
 'were',
 'latter',
 'will',
 'many',
 'see',
 'each',
 'fifty',
 'their',
 'mine',
 'amongst',
 'upon',
 'unless',
 'via',
 'forty',
 'are',
 'however',
 'yourself',
 'just',
 'hereafter',
 'those',
 'herself',
 'no',
 'to',
 "n't",
 'both',
 'everywhere',
 'alone',
 'rather',
 'yet',
 'last',
 'the',
 'nobody',
 'themselves',
 'six',
 'must',
 'does',
 'whither',
 'five',
 'then',
 'me',
 're',
 'anything',
 'out',
 'that',
 'thereby',
 'wherever',
 'over',
 'than',
 'formerly',
 'somehow',
 'therefore',
 'together',
 'much',
 'latterly',
 'anyone',
 'off',
 'cannot',
 '’m',
 'various',
 'several',
 'get',
 'throughout',
 '‘ll',
 'made',
 'beyond',
 'so',
 'been',
 'sometimes',
 'everyone',
 'or',
 'below',
 'yourselves',
 'neither',
 'whose',
 'ever',
 'become',
 'thru',
 'such',
 'using',
 'whole',
 'can',
 'how',
 'take',
 'be',
 'other',
 'put',
 'us',
 'during',
 'almost',
 'mostly',
 'one',
 'with',
 'n’t',
 'now

In [35]:
# for en_core_web_sm
!python3 -m spacy download en_core_web_sm

     |████████████████████████████████| 13.9 MB 4.6 MB/s eta 0:00:01
You should consider upgrading via the '/opt/anaconda3/bin/python3 -m pip install --upgrade pip' command.
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


In [36]:
import spacy 

nlp = spacy.load('en_core_web_sm')

def text_data_cleaning(sentence):
    
    doc = nlp(sentence)
    
    
    tokens = [] # list of tokens
    for token in doc:
        if token.lemma_ != "-PRON-":
            temp = token.lemma_.lower().strip()
        else:
            temp = token.lower_
        tokens.append(temp)
    
    cleaned_tokens = []
    for token in tokens:
        if token not in stopwords and token not in punct:
            cleaned_tokens.append(token)
    return cleaned_tokens

In [37]:
# if root form of that word is not pronoun then it is going to convert that into lower form
# and if that word is a proper noun, then we are directly taking lower form, because there is no lemma for proper noun

In [38]:
# stopwords and punctuations removed
text_data_cleaning("Hello all, It's a beautiful day outside there!")

['hello', 'beautiful', 'day', 'outside']

# Vectorization Feature Engineering (TF-IDF)

In [39]:
from sklearn.svm import LinearSVC
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.pipeline import Pipeline

# tokenizer=text_data_cleaning, tokenization will be done according to this function
tfidf = TfidfVectorizer(tokenizer=text_data_cleaning)

classifier = LinearSVC()

# Train the model

### Splitting the dataset into the Train and Test set

In [40]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

In [41]:
# 2198 samples in training dataset and 550 in test dataset
x_train.shape, x_test.shape

((2198,), (550,))

In [42]:
x_train.head()

2572    An Italian reviewer called this "a small, grea...
526                          And it was way to expensive.
1509    As an earlier review noted, plug in this charg...
144     Nice blanket of moz over top but i feel like t...
2483    The film gives meaning to the phrase, "Never i...
Name: Review, dtype: object

# Fit 

In [43]:
# first do vectorization then do classification
clf = Pipeline([('tfidf',tfidf), ('clf',classifier)])

clf.fit(x_train, y_train)

Pipeline(steps=[('tfidf',
                 TfidfVectorizer(tokenizer=<function text_data_cleaning at 0x7fce8ad4eee0>)),
                ('clf', LinearSVC())])

# Predict 

In [44]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

y_pred = clf.predict(x_test)

# confusion_matrix
confusion_matrix(y_test, y_pred)

array([[202,  77],
       [ 47, 224]])

In [45]:
# classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.81      0.72      0.77       279
           1       0.74      0.83      0.78       271

    accuracy                           0.77       550
   macro avg       0.78      0.78      0.77       550
weighted avg       0.78      0.77      0.77       550



In [46]:
accuracy_score(y_test, y_pred)

0.7745454545454545

# Testing

In [47]:
clf.predict(["Wow, I am learning Natural Language Processing in fun fashion!"])

array([1])

> output is 1, that means review is positive

In [48]:
clf.predict(["It's hard to learn new things!"])

array([0])

> output is 0, that means review is Negative